runtime/src/main/scala/akka/grpc/internal/ClientState.scala

```diff
@@ -94,25 +100,40 @@ final class ClientState(settings: GrpcClientSettings, channelFactory: GrpcClient
     }
   }
 
-  private def create(): InternalChannel = {
-    val internalChannel: InternalChannel = channelFactory(settings)
-    internalChannel.done.onComplete {
-      case Failure(_: ClientConnectionException | _: NoTargetException) =>
-        val old = internalChannelRef.get()
-        if (old.isDefined) {
-          val newInternalChannel = create()
-          // Only one client is alive at a time. However a close() could have happened between the get() and this set
-          if (!internalChannelRef.compareAndSet(old, Some(newInternalChannel))) {
-            // close the newly created client we've been shutdown
-            ChannelUtils.close(newInternalChannel)
-          }
-        }
-      case Failure(_) =>
+  private def create(): Future[InternalChannel] =
+    Patterns.retry(
+      () => channelFactory(settings),
+      settings.creationAttempts,
+      settings.creationDelay,
+      // TODO #733 remove cast once we update Akka
+      mat.asInstanceOf[ActorMaterializer].system.scheduler,
+      mat.asInstanceOf[ActorMaterializer].system.dispatcher)
+
+  private def recreateOnFailure(done: Future[Done]): Unit =
+    done.onComplete {
+      case Failure(e: ClientConnectionException) =>
+        log.warning(s"Transient connection error [${e.getMessage}], recreating client")
+        recreate()
+      case Failure(e) =>
+        // TODO This makes the client unusable. Perhaps we should retry here too?
+        log.warning(s"Unrecoverable connection error [${e.getMessage}], closing client")
```

Comment:
Yeah, maybe retry for all exceptions instead and see if that has any surprising consequences?